File: 프로그래머스/lv1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.py
def solution(numbers, hand):
    keypad = {
        1: [0,0],
        2: [0,1],
        3: [0,2],
        4: [1,0],
        5: [1,1],
        6: [1,2],
        7: [2,0],
        8: [2,1],
        9: [2,2],
        0: [3,1]
    }
    left = [3,0]
    right = [3,2]
    
    result = ''
    for num in numbers:
        a, b = keypad[num]
        if num in [1,4,7]:
            result += "L"
            left = [a,b]
            continue
        elif num in [3,6,9]:
            result += "R"
            right = [a,b]
            continue
        
        l = abs(left[0] - a) + abs(left[1] - b)
        r = abs(right[0] - a) + abs(right[1] - b)
        print(l, r)
        if l < r: 
            result += "L"
            left = [a,b]
        elif r < l:
            result += "R"
            right = [a,b]
        elif r == l:
            if hand == "right":
                result += "R"
                right = [a,b]
            else:
                result += "L"
                left = [a,b]
        
        
    return result

File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.java
import java.util.Map;

class Solution {
    public String solution(int[] numbers, String hand) {
        int[] l = {3, 0};
        int[] r = {3, 2};

        Map<Integer, int[]> dic = Map.of(
                1, new int[]{0, 0},
                2, new int[]{0, 1},
                3, new int[]{0, 2},
                4, new int[]{1, 0},
                5, new int[]{1, 1},
                6, new int[]{1, 2},
                7, new int[]{2, 0},
                8, new int[]{2, 1},
                9, new int[]{2, 2},
                0, new int[]{3, 1}
        );

        StringBuilder sb = new StringBuilder();
        
        for (int number: numbers) {
            if (number == 1 || number == 4 || number == 7) {
                sb.append("L");
                l = dic.get(number);
            } else if (number == 3 || number == 6 || number == 9) {
                sb.append("R");
                r = dic.get(number);
            } else {
                int[] target = dic.get(number);
                int lDist = Math.abs(l[0] - target[0]) + Math.abs(l[1] - target[1]);
                int rDist = Math.abs(r[0] - target[0]) + Math.abs(r[1] - target[1]);
                
                if (lDist < rDist || (lDist == rDist && hand.equals("left"))) {
                    sb.append("L");
                    l = target;
                } else {
                    sb.append("R");
                    r = target;
                }
            }
        }
        
        return sb.toString();
    }
}